### PR TITLE
Deprecate relying on equals for up-to-date checks

### PR DESCRIPTION
**Pull request by wolfs**
As a preparation to solve https://github.com/gradle/gradle/issues/17324 in Gradle 8.0.

### CODE REVIEW

File: subprojects/model-core/src/main/java/org/gradle/internal/snapshot/impl/SerializedValueSnapshot.java [hunk 1]
```diff
@@ -73,6 +74,12 @@ private boolean hasSameSerializedValue(Object value, ValueSnapshot snapshot) {
             // Deserialize the old value and use the equals() implementation. This will be removed at some point
             Object oldValue = populateClass(value.getClass());
             if (oldValue.equals(value)) {
+                DeprecationLogger.deprecateIndirectUsage("Using objects as inputs which have a different serialized form but are equals")
+                    .withContext("Type '" + value.getClass().getName() + "' has a custom equals implementation.")
+                    .withAdvice("Use @Nested instead.")
```
Comment (lptr) [inline subprojects/model-core/src/main/java/org/gradle/internal/snapshot/impl/SerializedValueSnapshot.java:79]: I think we should give a little bit more explanation here about how using `@Nested` solve this. How about:

```suggestion
                    .withAdvice("Declare the property as @Nested instead to expose its properties as inputs.")
```

**Comment by wolfs** [comment]
@bot-gradle test QFL

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by wolfs** [comment]
@bot-gradle test and merge

**Pull request merged**